packages/medusa/src/services/cart.ts

```diff
@@ -1801,6 +1807,98 @@ class CartService extends TransactionBaseService {
     )
   }
 
+  async upsertPaymentSessions(cartOrId: Cart | string) {
+    return this.atomicPhase_(async (manager) => {
+      const cart =
+        typeof cartOrId !== `string`
+          ? cartOrId
+          : await this.retrieveWithTotals(
+              cartOrId,
+              {
+                relations: [
+                  "items.variant.product.profiles",
+                  "items.adjustments",
+                  "discounts.rule",
+                  "gift_cards",
+                  "shipping_methods.shipping_option",
+                  "billing_address",
+                  "shipping_address",
+                  "region.tax_rates",
+                  "region.payment_providers",
+                  "payment_sessions",
+                  "customer",
+                ],
+              },
+              { force_taxes: true }
+            )
+
+      const psRepo = manager.withRepository(this.paymentSessionRepository_)
+
+      const paymentProviderServiceTx =
+        this.paymentProviderService_.withTransaction(manager)
+
+      const { total } = cart
+
+      await Promise.all(
+        cart.payment_sessions.map(async (session) => {
+          /**
+           * Update selected and initiated payment session with the latest cart data.
+           */
+          if (session.is_selected && session.is_initiated) {
+            const updateData = {
+              cart,
+              customer: cart.customer,
+              amount: cart.total!,
+              currency_code: cart.region.currency_code,
+              provider_id: session.provider_id,
+            }
+
+            return paymentProviderServiceTx.updateSession(session, updateData)
+          }
+
+          /**
+           * Upsert payment sessions that are not selected with the latest cart data.
+           *
+           * Initiated sessions are deleted remotely and created anew locally.
+           * Uninitiated session are updated locally.
+           */
+          let upsertedSession: PaymentSession
+
+          if (session.is_initiated) {
+            await paymentProviderServiceTx.deleteSession(session)
+            return psRepo.create({
+              cart_id: cart.id,
+              data: {},
+              status: PaymentSessionStatus.PENDING,
+              amount: total,
+              is_initiated: false,
+              provider_id: session.provider_id,
+            })
+          } else {
+            upsertedSession = { ...session, amount: total } as PaymentSession
+          }
+
+          return psRepo.save(upsertedSession)
+        })
+      )
+    })
+  }
+
+  async deletePaymentSessionLocalAndRemote(session: PaymentSession) {
```

Comment:
thought: again this feels like it is beyond the scope of what CartService should do.

Maybe we could move this logic to the PaymentCollection service and slowly migrate management of the session<->cart relationship there as we build more capabilities into workflows.

Reply:
I've had difficulty deciding if this refactoring should be part of building the workflow and ultimately ended up not doing it.

My reasons were:
1. It would significantly widen the scope of the workflow PR(s)
2. It would involve refactoring a lot of business logic that would _possibly_ need to be refactored again when we modularize the domains e.g. payments + carts. 
3. It would entail a lot of work (just) adding types for all the refactored bits because we don't want to depend on `@medusajs/medusa` in `@medusajs/types`. Types that would initially be almost 1:1 with our Typeorm models, i.e. types that are subject to change when we modularize and switch to MikroORM.
4. It would introduce a lot of business logic that is not battle-tested. However, this is not really a concern as we will not be announcing any stability around workflows any time soon.

I think these can be boiled down to a main concern, which is: Is it worth doing this now, given that it will likely change in the near future? With the current state of the PR, we at least know which units of logic need to be refactored by looking at the workflow steps.

All that said, I think you raise a very valid point, so I am keen to give this a shot if we are all aligned on this. Successful or not, it would give us a good understanding of what is required to move this around.

What do you think?